rust/feature-flags/src/flag_matching.rs

```diff
@@ -805,6 +824,85 @@ impl FeatureFlagMatcher {
         }
     }
 
+    pub async fn evaluate_cohort_filters(
+        &self,
+        filters: &[PropertyFilter],
+        target_properties: &HashMap<String, Value>,
+    ) -> Result<bool, FlagError> {
+        Box::pin(self.evaluate_potentially_nested_cohort_filters(filters, target_properties)).await
+    }
+
+    async fn evaluate_potentially_nested_cohort_filters(
+        &self,
+        filters: &[PropertyFilter],
+        target_properties: &HashMap<String, Value>,
+    ) -> Result<bool, FlagError> {
+        let mut cohort_filters = Vec::new();
+        let mut non_cohort_filters = Vec::new();
+
+        // Separate cohort filters from non-cohort filters
+        for filter in filters {
+            if filter.prop_type == "cohort" {
+                cohort_filters.push(filter);
+            } else {
+                non_cohort_filters.push(filter);
+            }
+        }
+
+        // Evaluate non-cohort filters
+        for filter in &non_cohort_filters {
+            if !match_property(filter, target_properties, false).unwrap_or(false) {
+                return Ok(false);
+            }
+        }
+
+        // Evaluate cohort filters
+        if !cohort_filters.is_empty() {
+            let cohort_ids: HashSet<CohortId> = cohort_filters
+                .iter()
+                .filter_map(|f| f.value.as_i64().map(|id| id as CohortId))
+                .collect();
+
+            let cohorts = self.fetch_cohorts(cohort_ids.clone()).await?;
+            let seen_cohorts_cache: HashMap<CohortId, CohortOrEmpty> = cohorts
+                .into_iter()
+                .map(|cohort| (cohort.id, CohortOrEmpty::Cohort(cohort)))
+                .collect();
+
+            let sorted_cohort_ids = sort_cohorts_topologically(cohort_ids, &seen_cohorts_cache);
+
+            for cohort_id in sorted_cohort_ids {
+                if let Some(CohortOrEmpty::Cohort(cohort)) = seen_cohorts_cache.get(&cohort_id) {
+                    let cohort_property_filters = cohort.parse_filters()?; // TODO error handle
+                    let cohort_match = self
+                        .evaluate_cohort_filters(&cohort_property_filters, target_properties)
+                        .await?;
```

Comment:
then we recurse and do the same thing for every cohort that this depends on.

Reply:
Rather than recursing here, I think you could (roughly):
- Get _all_ the `PropertyFilter`s from all the cohorts you just fetched
- Split them as you did above into non-cohort and cohort filters
- Evaluate /all/ of the non-cohort filters, and early exit if any fail
- Push all the id's from the cohort filters into the `cohort_filters` vec
- loop on line 860.

This would have the advantage of being breadth-first rather than depth-first, which might let you early exit with fewer DB queries, and it also lets you avoid recursion.